openhtf/core/test_state.py

```diff
@@ -251,16 +250,26 @@ def finalize_normally(self):
     if self._is_aborted():
       return
 
-    allowed_outcomes = {measurements.Outcome.PASS}
-    if conf.allow_unset_measurements:
-      allowed_outcomes.add(measurements.Outcome.UNSET)
+    all_skipped = bool(self.test_record.phases)
+    any_failed = False
 
-    if any(meas.outcome not in allowed_outcomes
-           for phase in self.test_record.phases
-           for meas in phase.measurements.itervalues()):
+    for phase in self.test_record.phases:
```

Comment:
it's a bit less efficient (linearly, though), but the number of phases should be small enough that i think the more pythonic version is more readable:
```
  if not self.test_record.phases:
    # Vacuously PASS a TestRecord with no phases.
    self._finalize(test_record.Outcome.PASS)
  elif any(phase.outcome == test_record.PhaseOutcome.FAIL for phase in self.test_record.phases):
    # Any FAIL phase results in a test failure.
    self._finalize(test_record.Outcome.FAIL)
  elif all(phase.outcome == test_record.PhaseOutcome.SKIP for phase in self.test_record.phases):
    # Make sure at least one phase ran (wasn't skipped).
    # ...
  else:
    # Otherwise all phases that ran passed, test is a PASS.
    self._finalize(test_record.Outcome.PASS)
```